### ISSUE TITLE
`--force-date` option doesn't work

### ISSUE THREAD

**Issue by mattdibi** [PentesterES/Delorean#7]
As per title, the `--force-date` doesn't work properly.

### Testing configuration

Delorean running on my Mac with:

```bash
➜  Delorean git:(master) $ sudo ./delorean.py -d "2022-04-01 15:30" -n
Password:
[20:57:20] Sent to 127.0.0.1:49233 - Going to the future! 2022-04-01 13:30
[20:57:22] Sent to 127.0.0.1:57031 - Going to the future! 2022-04-01 13:30
[20:57:50] Sent to 192.168.1.122:56136 - Going to the future! 2022-04-01 13:30
```

Client running chrony with the following configuration:

```conf
# Use public NTP servers from the pool.ntp.org project.
server 192.168.1.51

# Record the rate at which the system clock gains/losses time.
driftfile /var/lib/chrony/drift

# Allow the system clock to be stepped in the first three updates
# if its offset is larger than 1 second.
makestep 1.0 3

# Enable kernel synchronization of the real-time clock (RTC).
rtcsync
```

Testing application `get_ntp.py`:

```python
import ntplib
from time import ctime
c = ntplib.NTPClient()
response = c.request('192.168.1.51')
print(ctime(response.tx_time))
```

### Default behaviour

Chrony refuses to synch with Delorean seemingly because **the time returned by Delorean doesn't changes** when used with the `--force-date` option.

```bash
jetson@jetson-desktop:~$ chronyc sources
210 Number of sources = 1
MS Name/IP address         Stratum Poll Reach LastRx Last sample
===============================================================================
^? 192.168.1.51                  2   6     1    21   -9632m[ -9632m] +/-   59ms
```

```bash
jetson@jetson-desktop:~$ date
ven 25 mar 2022, 22.01.08, CET
```

```bash
jetson@jetson-desktop:~$ python3 get_ntp.py
Fri Apr  1 15:30:00 2022
jetson@jetson-desktop:~$ python3 get_ntp.py
Fri Apr  1 15:30:00 2022
jetson@jetson-desktop:~$ python3 get_ntp.py
Fri Apr  1 15:30:00 2022
jetson@jetson-desktop:~$ python3 get_ntp.py
Fri Apr  1 15:30:00 2022
```

### Patched behaviour

Using the following hotfix

```diff
diff --git a/delorean.py b/delorean.py
index a4d9873..0e513aa 100755
--- a/delorean.py
+++ b/delorean.py
@@ -43,6 +43,7 @@ class NTProxy(threading.Thread):
     forced_step = float(0)
     forced_date = float(0)
     forced_random = False
+    initial_timestamp = time.time()
     # Temporal control
     seen = {}
     # Constructor
@@ -152,7 +153,12 @@ class NTProxy(threading.Thread):
             try:
                 data, source = self.socket.recvfrom(1024)
                 info = self.extract(data)
+
                 timestamp = self.newtime(info["tx_timestamp"] - self.ntp_delta)
+                timestamp += time.time() - self.initial_timestamp
+
                 fingerprint, data = self.response(info, timestamp)
                 if self.skim_step != 0:
                     for t in range(0, 10):
```

```bash
jetson@jetson-desktop:~$ python3 get_ntp.py
Fri Apr  1 15:30:02 2022
jetson@jetson-desktop:~$ python3 get_ntp.py
Fri Apr  1 15:30:04 2022
jetson@jetson-desktop:~$ python3 get_ntp.py
Fri Apr  1 15:30:05 2022
jetson@jetson-desktop:~$ python3 get_ntp.py
Fri Apr  1 15:30:06 2022
```

it works as expected

```bash
jetson@jetson-desktop:~$ chronyc sources
210 Number of sources = 1
MS Name/IP address         Stratum Poll Reach LastRx Last sample
===============================================================================
^* 192.168.1.51                  2   6     7    14    -24ms[ -2692h] +/-   56ms
```

```bash
jetson@jetson-desktop:~$ date
ven  1 apr 2022, 15.35.03, CEST
```

### Conclusions

The problem seems to be related to the fact that with the default behaviour of the `--force-date` the time returned by Delorean is always the same.

**Comment by jselvi** [comment]
Hey @mattdibi, thanks for taking the time to write this report and the fix.

Actually, `--force-date` flag was supposed to work as you have seen, not as you thought. If you want the date to be updated each time someone requests a date (so `--force-date` only sets the initial date), that's a relative update, and it's implemented with the `--force-step` flag.

Have you tried to use something like `$ sudo ./delorean.py -d 2w -n` ? That may be what you are looking for.
I'm happy to accept a contribution if you think `--force-step` to support using an absolute time and convert it to a relative one under the hood, if there is any scenario where this could be useful, but I prefer to keep `--force-date` as it was originally.

**Comment by mattdibi** [comment]
Hi @jselvi, thanks for the feedback.

If it's intended behavior I don't think my patch makes much sense then.

> Have you tried to use something like $ sudo ./delorean.py -d 2w -n ?

I think you meant `$ sudo ./delorean.py -s 2w -n`, if so yeah I did and it works flawlessly. I expected, maybe wrongfully, that the `--force-step` and `--force-date` options were just two different way to express the target date/time but shared the same behavior. I think all of this can be addressed with a note in the documentation.

Just a final question then: what's the purpose of the `--force-date` option if it doesn't cheat NTP servers? By not updating the reported date/time as time passes some servers (in this case chrony) don't trust it as a time source. I ask because I'm not knowledgeable in this field (I just used Delorean for QA testing)

**Comment by jselvi** [comment]
Hey @mattdibi, yes, I meant "-s". It was a copy & paste error :)

I have added a clarification for this in the README file: https://github.com/PentesterES/Delorean/commit/0291151fc16fc91542690ca81cbe61ecdb2804fa

The reason this attack was implemented is because I implemented a number of attacks that came to my mind to test NTP configurations. Some of them resulted to be useful, and some others don't. At some point I removed the attacks that were not effective at all (https://github.com/PentesterES/Delorean/commit/44a95cc787e61ffe9c468686072bbe1a97d67fd6). The `--force-date` option actually worked in some configurations, so I kept it.
